src/driver-sinowealth.c

```diff
@@ -0,0 +1,475 @@
+#include "libratbag-hidraw.h"
+
+#define SINOWEALTH_REPORT_ID_CONFIG 0x4
+#define SINOWEALTH_REPORT_ID_CMD 0x5
+#define SINOWEALTH_CMD_FIRMWARE_VERSION 0x1
+#define SINOWEALTH_CMD_GET_CONFIG 0x11
+#define SINOWEALTH_CONFIG_SIZE 520
+
+#define SINOWEALTH_XY_INDEPENDENT 0x80
+
+/* The PC software only goes down to 400, but the PMW3360 doesn't care */
+#define SINOWEALTH_DPI_MIN 100
+#define SINOWEALTH_DPI_MAX 12000
+#define SINOWEALTH_DPI_STEP 100
+
+/* other models might have up to eight */
+#define SINOWEALTH_NUM_DPIS 6
+
+#define SINOWEALTH_RGB_BRIGHTNESS_BITS 0xF0
+#define SINOWEALTH_RGB_SPEED_BITS 0x0F
+
+struct RGB8 {
+	uint8_t r, g, b;
+} __attribute__((packed));
+
+_Static_assert(sizeof(struct RGB8) == 3, "Invalid size");
+
+struct RBG8 {
+	uint8_t r, b, g;
+} __attribute__((packed));
+
+_Static_assert(sizeof(struct RBG8) == 3, "Invalid size");
+
+enum rgb_effect {
+	RGB_OFF = 0,
+	RGB_GLORIOUS = 0x1,   /* unicorn mode */
+	RGB_SINGLE = 0x2,     /* single constant color */
+	RGB_BREATHING = 0x5,  /* RGB breathing */
+	RGB_BREATHING7 = 0x3, /* breathing with seven colors */
+	RGB_BREATHING1 = 0xa, /* single color breathing */
+	RGB_TAIL = 0x4,       /* idk what this is supposed to be */
+	RGB_RAVE = 0x7,       /* ig */
+	RGB_WAVE = 0x9
+};
+
+struct sinowealth_config_report {
+	uint8_t report_id; /* SINOWEALTH_REPORT_ID_CONFIG */
+	uint8_t command_id;
+	uint8_t unk1;
+	uint8_t config_write;
+	/* always 0 when config is read from device,
+	 * has to be 0x7b when writing config to device
+	 */
+	uint8_t unk2[6];
+	uint8_t config;
+	/* 0x80 - SINOWEALTH_XY_INDEPENDENT */
+	uint8_t dpi_count:4;
+	uint8_t active_dpi:4;
+	uint8_t dpi_enabled;
+	/* bit set: disabled, unset: enabled
+	 * this structure has support for eight DPI slots,
+	 * but the glorious software only exposes six
+	 */
+	uint8_t dpi[16];
+	/* DPI/CPI is encoded in the way the PMW3360 sensor accepts it
+	 * value = (DPI - 100) / 100
+	 * If XY are identical, dpi[0-6] contain the sensitivities,
+	 * while in XY independent mode each entry takes two chars for X and Y.
+	 */
+	struct RGB8 dpi_color[8];
+	uint8_t rgb_effect;
+	/* see enum rgb_effect */
+	uint8_t glorious_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t glorious_direction;
+	uint8_t single_mode;
+	struct RBG8 single_color;
+	uint8_t breathing_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t breathing_colorcount;
+	/* 7, constant */
+	struct RBG8 breathing_colors[7];
+	uint8_t tail_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t unk4[33];
+	uint8_t rave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	struct RBG8 rave_colors[2];
+	uint8_t wave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	uint8_t breathing1_mode;
+	/* 0x1/2/3 - speed */
+	struct RBG8 breathing1_color;
+	uint8_t unk5;
+	uint8_t lift_off_distance;
+	/* 0x1 - 2 mm
+	 * 0x2 - 3 mm
+	 */
+	uint8_t padding[520 - 131];
```

Comment:
`SINOWEALTH_CONFIG_SIZE` instead of a literal 520 here.

Reply:
I added another #define for SINOWEALTH_CONFIG_SIZE_USED which made the code in read_profile code clearer as well.